core/src/sumeragi/main_loop.rs

```diff
@@ -130,28 +132,55 @@ impl Sumeragi {
     }
 
     #[allow(clippy::panic)]
-    fn receive_network_packet(&self, view_change_proof_chain: &mut ProofChain) -> Option<Message> {
-        let current_topology = &self.current_topology;
-        match self.message_receiver.try_recv() {
-            Ok(packet) => {
-                if let Err(error) = view_change_proof_chain.merge(
-                    packet.view_change_proofs,
-                    &current_topology.sorted_peers,
-                    current_topology.max_faults(),
-                    self.wsv.latest_block_hash(),
-                ) {
-                    trace!(%error, "Failed to add proofs into view change proof chain")
-                }
-                Some(packet.message)
-            }
-            Err(recv_error) => match recv_error {
-                mpsc::TryRecvError::Empty => None,
-                mpsc::TryRecvError::Disconnected => {
-                    panic!("Sumeragi message pump disconnected. This is not a recoverable error.")
-                    // TODO: Use early return.
-                }
-            },
+    fn receive_network_packet(
+        &self,
+        view_change_proof_chain: &mut ProofChain,
+        control_message_in_a_row_counter: &mut usize,
+    ) -> Option<Message> {
+        const MAX_CONTROL_MSG_IN_A_ROW: usize = 25;
+
+        let mut packet = if *control_message_in_a_row_counter < MAX_CONTROL_MSG_IN_A_ROW {
+            *control_message_in_a_row_counter += 1;
+            self.control_message_receiver
+                .try_recv()
+                .map_err(|recv_error| {
+                    assert!(
+                        recv_error != mpsc::TryRecvError::Disconnected,
+                        "Sumeragi control message pump disconnected. This is not a recoverable error."
+                    )
+                })
+                .ok()
+                .map(std::convert::Into::into)
+        } else {
+            None
+        };
+
+        if packet.is_none() {
+            *control_message_in_a_row_counter = 0;
+            packet = self
+                .message_receiver
+                .try_recv()
+                .map_err(|recv_error| {
+                    assert!(
+                        recv_error != mpsc::TryRecvError::Disconnected,
+                        "Sumeragi message pump disconnected. This is not a recoverable error."
+                    )
+                })
+                .ok();
         }
+
+        packet.map(|packet| {
+            if let Err(error) = view_change_proof_chain.merge(
+                packet.view_change_proofs,
+                &self.current_topology.sorted_peers,
+                self.current_topology.max_faults(),
+                self.wsv.latest_block_hash(),
+            ) {
+                trace!(%error, "Failed to add proofs into view change proof chain")
+            }
+
+            packet.message
+        })
```

Comment:
```suggestion
        packet.map_or_else(|| {
            *control_message_in_a_row_counter = 0;
            packet = self
                .message_receiver
                .try_recv()
                .map_err(|recv_error| {
                    assert!(
                        recv_error != mpsc::TryRecvError::Disconnected,
                        "Sumeragi message pump disconnected. This is not a recoverable error."
                    )
                })
                .ok();
        }, |packet| {
            if let Err(error) = view_change_proof_chain.merge(
                packet.view_change_proofs,
                &self.current_topology.sorted_peers,
                self.current_topology.max_faults(),
                self.wsv.latest_block_hash(),
            ) {
                trace!(%error, "Failed to add proofs into view change proof chain")
            }

            packet.message
        })
```

Reply:
I have taken this suggestion and gone a little further.